File: 474/474.py
# Leetcode 474: Ones and Zeroes
# https://leetcode.com/problems/ones-and-zeroes/
# Solved on 11th of November, 2025
class Solution:
    def findMaxForm(self, strs: list[str], m: int, n: int) -> int:
        """
        Finds the maximum number of strings that can be formed from the given array of strings,
        given the maximum number of '0's (m) and '1's (n) allowed.

        :param strs: A list of strings consisting of '0's and '1's.
        :param m: The maximum number of '0's allowed.
        :param n: The maximum number of '1's allowed.
        :return: The maximum number of strings that can be formed.
        """
        dp = [[0 for _ in range(n + 1)] for _ in range(m + 1)]

        for s in strs:
            numZeros = s.count('0')
            numOnes = s.count('1')

            for zeroBudget in range(m, numZeros - 1, -1):
                for oneBudget in range(n, numOnes - 1, -1):
                    dp[zeroBudget][oneBudget] = max(
                        dp[zeroBudget][oneBudget],
                        1 + dp[zeroBudget - numZeros][oneBudget - numOnes]
                    )

        return dp[m][n]

File: 474/474.java
// Leetcode 474: Ones and Zeroes
// https://leetcode.com/problems/ones-and-zeroes/
// Solved on 11th of November, 2025
class Solution {
    /**
     * Finds the maximum number of strings that can be formed from the given array `strs`
     * such that the total number of '0's in the chosen strings is at most `m`
     * and the total number of '1's is at most `n`.
     * @param strs An array of binary strings.
     * @param m The maximum allowed number of '0's.
     * @param n The maximum allowed number of '1's.
     * @return The maximum number of strings that can be formed.
     */
    public int findMaxForm(String[] strs, int m, int n) {
        int[][] dp = new int[m + 1][n + 1];
        for (String s : strs) {
            int zeros = 0;
            int ones = 0;
            for (int k = 0; k < s.length(); k++) {
                if (s.charAt(k) == '0') {
                    zeros++;
                } else {
                    ones++;
                }
            }
            if (zeros > m || ones > n) {
                continue;
            }
            for (int i = m; i >= zeros; i--) {
                for (int j = n; j >= ones; j--) {
                    int cand = dp[i - zeros][j - ones] + 1;
                    if (cand > dp[i][j]) {
                        dp[i][j] = cand;
                    }
                }
            }
        }
        return dp[m][n];
    }
}